api/envoy/config/rbac/v2alpha/rbac.proto

```diff
@@ -0,0 +1,172 @@
+//   suffix "/reviews" when request header "version" set to either "v1" or "v2".
+//
+//   action_type: ALLOW
+//   policies:
+//     "service-admin":
+//       permissions:
+//       -
+//       principals:
+//         authenticated:
+//           name: "cluster.local/ns/default/sa/admin"
+//     "product-viewer":
+//       permissions:
+//       - paths: [prefix: "/products", suffix: "/reviews"]
+//         methods: ["GET"]
+//         conditions:
+//         - header:
+//             key: "version"
+//             values: [simple: "v1", simple: "v2"]
+//       principals:
+//       -
+//
+message RBAC {
+  // Should we do white-list or black-list style access control.
+  enum ActionType {
+    // The policies grant access to principals. The rest is denied. This is white-list style
+    // access control. This is the default type.
+    ALLOW = 0;
+
+    // The policies deny access to principals. The rest is allowed. This is black-list style
+    // access control.
+    DENY = 1;
+  }
+
+  ActionType action_type = 1;
+
+  // Maps from policy name to policy.
+  map<string, Policy> policies = 2;
+}
+
+// Policy specifies a role and the principals that are assigned/denied the role.
+message Policy {
+ // Required. The set of permissions that define a role.
+ repeated Permission permissions = 1 [(validate.rules).repeated .min_items = 1];
+
+ // Required. List of principals that are assigned/denied the role based on “action_type”.
+ repeated Principal principals = 2 [(validate.rules).repeated .min_items = 1];
+}
+
+message StringMatch {
+  oneof match_pattern {
+    // The input string must match exactly the string specified here.
+    // Or it is a "*", which means that it matches any string.
+    string simple = 1;
+
+    // The input string must have the prefix specified here.
+    string prefix = 2;
+
+    // The input string must have the suffix specified here.
+    string suffix = 3;
+
+    // The input string must match the regular expression specified here.
+    // The regex grammar is defined `here
+    // <http://en.cppreference.com/w/cpp/regex/ecmascript>`_.
+    string regex = 4;
+ }
+}
+
+// Specifies how to match an entry in a map.
+message MapEntryMatch {
+  // The key to select an entry from the map.
+  string key = 1;
+
+  // A list of matched values.
+  repeated StringMatch values = 2;
+}
+
+// Specifies how to match IP addresses.
+message IpMatch {
+  // IP addresses in CIDR notation.
+  repeated envoy.api.v2.core.CidrRange cidrs = 1;
+}
+
+// Specifies how to match ports.
+message PortMatch {
+  // Port numbers.
+  repeated uint32 ports = 1;
+}
+
+// Permission defines a permission to access the service.
+message Permission {
+  // Optional. A list of HTTP paths or gRPC methods.
+  // gRPC methods must be presented as fully-qualified name in the form of
+  // packageName.serviceName/methodName.
+  // If this field is unset, it applies to any path.
+  repeated StringMatch paths = 1;
```

Comment:
Is there any interest on matching on query params out of curiosity (I don't have a concrete use case)?

Reply:
Let's add it later when someone have use case?

Reply:
Good question. We may need to match on query parameters. But the [current definition of query parameters in AttributeContext](https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/auth/v2alpha/attribute_context.proto#L90) is:
```
    // The HTTP URL query in the format of `name1=value`&name2=value2`, as it
    // appears in the first line of the HTTP request. No decoding is performed.
    string query = 7;
```
In order to refer to a particular query parameter (say, "api_key"), we have to parse the query string. @wora @mandarjog Can we make the query parameter type to be `map<string, string>`?

Reply:
+1 to change to map, but I per @lizan I would add this in a follow up.

Reply:
I will leave it in a follow up PR.